design/nfs.md

```diff
@@ -0,0 +1,157 @@
+NFS supports multiple read/write simultaneously so a single share can be attached to multiple pods.
+
+## Design
+
+Rook brings storage platforms to Kubernetes with a fully Kubernetes-native design. It integrates deeply into cloud native environments leveraging extension points and providing a seamless experience for scheduling, lifecycle management, resource management, security, monitoring, and user experience.
+With this project Rook is providing another widely adopted storage option for admins and users of cloud-native environments. This project will automate NFS starting from its configuration (such as allowed hosts, read/write permissions etc.) to deployment and provisioning. The operations on NFS which cannot be done natively by Kubernetes will be automated.
+NFS doesn’t provide an internal provisioner for kubernetes, so Rook is needed as an external provisioner.
+
+### Initial Setup
+
+The flow of creating NFS backed storage in Rook is
+1. The settings are determined and saved in an NFS server CRD (rook-nfs.yaml)
+2. `kubectl create -f rook-nfs.yaml`
+3. Rook is notified of the new CRD
+   - Rook starts the NFS daemon on the desired nodes in the cluster
+   - Rook creates a storage class
+4. The pod which requires NFS storage creates a PVC to consume the volume.
+
+### Deployed service
+
+When the NFS CRD instance is created, Rook responds to this request by starting the NFS daemon with the configuration and exports stated in the CRD and creates a service to expose NFS.
+
+### NFS CRD
+
+The NFS CRD spec will specify the following:
+1. NFS server configuration
+    - export (The directory being exported)
+    - host (The host or network to which the export is being shared)
+    - options (The options to be used for the host) e.g., read and write permission, sync, root_squash.
+2. NFS server storage backend configuration. E.g., configuration for various storage backends(ceph, ebs, azure disk etc) that will be shared using NFS.
+
+The parameters to configure NFS are:
+
+| Parameter                       | Description                                                              | Default              |
+|---------------------------------|--------------------------------------------------------------------------|----------------------|
+| `replicas`                      | The no. of NFS daemon to start                                            | `1`                  |
+| `serverConfig`                  | NFS server configuration parameters                                      | <none>               |
+| `serverConfig.validHosts`       | The host or network to which export is being shared.                     | Current pod network  |
+| `serverConfig.accessMode`       | Reading and Writing permissions on the share.                             | `RO`                 |
```

Comment:
What values are possible? Would be nice for the CRD to accept readable values such as `ReadOnly` or `ReadWrite`.

Reply:
You need to make squash, and access mode as per client option.
A client IP can have RW and No_Squash access to an export, while an another IP can have RO and Root_Squash access to the same export.

See,
https://github.com/nfs-ganesha/nfs-ganesha/blob/next/src/config_samples/export.txt